File: 018.zig
const std = @import("std");
const stdout = std.io.getStdOut().writer();
const data = @embedFile("./0018_triangle.txt");
const common = @import("common");

fn solve(triangle: [][]u16) u64 {
    var y: u64 = triangle.len - 2;
    while (y > 0) : (y -= 1) {
        for (0..y + 1) |x| {
            triangle[y][x] = @max(triangle[y][x] + triangle[y + 1][x], triangle[y][x] + triangle[y + 1][x + 1]);
        }
    }

    triangle[0][0] = @max(triangle[0][0] + triangle[1][0], triangle[0][0] + triangle[1][1]);
    return triangle[0][0];
}

pub fn main() !void {
    var allocator = std.heap.page_allocator;

    var triangle = try common.load_triangle(&allocator, data);
    try stdout.print("{}\n", .{solve(triangle)});
    common.unload_triangle(triangle, &allocator);
}

File: 018.c
// 1074
#include <stdio.h>

#define EULER_COMMON_IMPLEMENTATION
#include "common.h"

int max(int a, int b)
{
    return (a > b)? a : b;
}

int triangle_at(int *triangle, int x, int y, int width, int height)
{
    if (x >= 0 && x < width && y >= 0 && y < height)
        return triangle[y*width + x];

    return -1;
}

int main()
{
    int width, height;
    int *triangle = load_triangle("./resources/0018_triangle.txt", &width, &height);

    int y = height - 2;
    int x = 0;
    while (y > 0) {
        for (int x = 0; x <= y; ++x) {
            triangle[y*width + x] = max(triangle_at(triangle, x, y, width, height) + triangle_at(triangle, x, y + 1, width, height),
                                        triangle_at(triangle, x, y, width, height) + triangle_at(triangle, x + 1, y + 1, width, height));
        }

        --y;
    }
    triangle[0] = max(triangle_at(triangle, 0, 0, width, height) + triangle_at(triangle, 0, 1, width, height),
                      triangle_at(triangle, 0, 0, width, height) + triangle_at(triangle, 1, 1, width, height));
    printf("%d\n", triangle[0]);

    return 0;
}
